geo/src/algorithm/lines_iter.rs

```diff
@@ -0,0 +1,328 @@
+use crate::{
+    CoordNum, Coordinate, Line, LineString, MultiLineString, MultiPolygon, Polygon, Rect, Triangle,
+};
+use core::slice;
+use std::fmt::Debug;
+use std::iter;
+
+/// Iterate over lines of a geometry.
+pub trait LinesIter<'a> {
+    type Scalar: CoordNum;
+    type Iter: Iterator<Item = Line<Self::Scalar>>;
+
+    /// Iterate over all exterior and (if any) interior lines of a geometry.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use geo::{Coordinate, Line};
+    /// use geo::line_string;
+    /// use geo::lines_iter::LinesIter;
+    ///
+    /// let ls = line_string![
+    ///     (x: 1., y: 2.),
+    ///     (x: 23., y: 82.),
+    ///     (x: -1., y: 0.),
+    /// ];
+    ///
+    /// let mut iter = ls.lines_iter();
+    /// assert_eq!(Some(Line::new(Coordinate{x: 1., y: 2.}, Coordinate{x: 23., y:82.})), iter.next());
+    /// assert_eq!(Some(Line::new(Coordinate{x: 23., y: 82.}, Coordinate{x: -1., y:0.})), iter.next());
+    /// assert_eq!(None, iter.next());
+    /// ```
+    fn lines_iter(&'a self) -> Self::Iter;
+}
+
+// ┌────────────────────────────┐
+// │ Implementation for Line    │
+// └────────────────────────────┘
+
+impl<'a, T: CoordNum + 'a> LinesIter<'a> for Line<T> {
+    type Scalar = T;
+    type Iter = iter::Copied<iter::Once<&'a Line<Self::Scalar>>>;
+
+    fn lines_iter(&'a self) -> Self::Iter {
+        iter::once(self).copied()
+    }
+}
+
+// ┌──────────────────────────────────┐
+// │ Implementation for LineString    │
+// └──────────────────────────────────┘
+
+impl<'a, T: CoordNum + 'a> LinesIter<'a> for LineString<T> {
+    type Scalar = T;
+    type Iter = LineStringIter<'a, Self::Scalar>;
+
+    fn lines_iter(&'a self) -> Self::Iter {
+        LineStringIter::new(self)
+    }
+}
+
+#[doc(hidden)]
+#[derive(Debug)]
+pub struct LineStringIter<'a, T: CoordNum>(slice::Windows<'a, Coordinate<T>>);
+
+impl<'a, T: CoordNum> LineStringIter<'a, T> {
+    fn new(line_string: &'a LineString<T>) -> Self {
+        Self(line_string.0.windows(2))
+    }
+}
+
+impl<'a, T: CoordNum> Iterator for LineStringIter<'a, T> {
+    type Item = Line<T>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // Can't use LineString::lines() because it returns an `impl Trait`
+        // and there is no way to name that type in `LinesIter::Iter` until [RFC 2071] is stabilized.
+        //
+        // [RFC 2071]: https://rust-lang.github.io/rfcs/2071-impl-trait-existential-types.html
+        self.0.next().map(|w| {
+            // slice::windows(2) is guaranteed to yield a slice with exactly 2 elements
+            unsafe { Line::new(*w.get_unchecked(0), *w.get_unchecked(1)) }
+        })
+    }
+}
+
+// ┌───────────────────────────────────────┐
+// │ Implementation for MultiLineString    │
+// └───────────────────────────────────────┘
+
+type MultiLineStringIter<'a, T> =
+    iter::Flatten<MapLinesIter<'a, slice::Iter<'a, LineString<T>>, LineString<T>>>;
```

Comment:
I would actually like to introduce newtypes for all these concrete `LinesIter::Iter` types because they currently leak implementation details (about how iterators are chained etc).

I followed the pattern in `CoordsIter` for this PR, but if there agreement, I'm happy to introduce newtypes for both `CoordsIter` and `LinesIter`.